File: Applied Course/5.Advanced Data Structure and Algorithms/4.Friend Circles.cpp
#include<bits/stdc++.h>
using namespace std;
// Logic-1
class Solution {
public:
    void DFS(vector<vector<int>>& M,int i,int n,vector<bool>& visited)
    {
        visited[i]=true;
        for(int j=0;j<n;j++)
            if(M[i][j]==1 && visited[j]==false)
                DFS(M,j,n,visited);
    }                
    
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n=isConnected.size();
        vector<bool> visited(n,false);
        int result=0;
        for(int i=0;i<n;i++)
        {
            if(visited[i]==false)
            {
                DFS(isConnected,i,n,visited);
                result++;
            }
        }
        return result;
    }
};
// Logic-2
class Solution {
public:
    int find(vector<int>& parent,int i)
    {
        if(parent[i]==-1)
            return i;
        return find(parent,parent[i]);
    }
    
    void Union(vector<int>& parent,int x,int y)
    {
        int xset=find(parent,x);
        int yset=find(parent,y);
        if(xset!=yset)
            parent[yset]=xset;
    }
    
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n=isConnected.size();
        vector<int> parent(n,-1);
        for(int i=0;i<n;i++)
            for(int j=0;j<n;j++)
                if(isConnected[i][j]==1 && i!=j)
                    Union(parent,i,j);
        int result=0;
        for(int i=0;i<n;i++)
            if(parent[i]==-1)
                result++;
        return result;
    }
};
int main()
{
    // https://leetcode.com/problems/number-of-provinces
    return 0;
}

File: Applied Course/5.Advanced Data Structure and Algorithms/4.Friend Circles.py
# https://leetcode.com/problems/number-of-provinces
# Logic-1
class Solution:
    def dfs(self,M,i,n,visited):
        visited[i]=True
        for j in range(n):
            if M[i][j]==1 and visited[j]==False:
                self.dfs(M,j,n,visited)
    
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n=len(isConnected)
        visited=[False for i in range(n)]
        result=0
        for i in range(n):
            if visited[i]==False:
                self.dfs(isConnected,i,n,visited)
                result+=1
        return result
# Logic-2
class Solution:
    def find(self,parent,i):
        if parent[i]==-1:
            return i
        return self.find(parent,parent[i])
    
    def union(self,parent,x,y):
        xset=self.find(parent,x)
        yset=self.find(parent,y)
        if xset!=yset:
            parent[yset]=xset
    
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n=len(isConnected)
        parent=[-1 for i in range(n)]
        for i in range(n):
            for j in range(n):
                if isConnected[i][j]==1 and i!=j:
                    self.union(parent,i,j)
        result=0
        for i in range(n):
            if parent[i]==-1:
                result+=1
        return result